man/man1/amk_grf.1
.\" Text automatically generated by txt2man
.TH amk_grf 1 "16 June 2026" "" "Scotch user's manual"
.SH NAME
\fBamk_grf \fP- create target architecture from source graph
\fB
.SH SYNOPSIS
.nf
.fam C
\fBamk_grf\fP [\fIoptions\fP] [\fIgfile\fP] [\fItfile\fP]

.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
The \fBamk_grf\fP program builds a decomposition-defined target
architecture \fItfile\fP from a source graph \fIgfile\fP.
.PP
Target architectures define the topology of the target graphs used
by static mapping programs \fBgmap\fP(1) and \fBdgmap\fP(1). Target
architectures can be either algorithmically-defined, for common,
regular topologies, or decomposition-defined, such as the ones
produced by \fBamk_grf\fP.
.PP
When the proper libraries have been included at compile time, \fBamk_grf\fP
can directly handle compressed files, both as input and output. A
stream is treated as compressed whenever its name is postfixed with
a compressed file extension, such as in 'brol.tgt.bz2' or '-.gz'. The
compression formats which can be supported are the bzip2 format
('.bz2'), the gzip format ('.gz'), and the lzma format ('.lzma').
.PP
Since decomposition-defined target architecture files have a size
which is quadratic in the number of target vertices, because of the
presence of a distance matrix structure, using compressed files to
store them may save a lot of space.
.SH OPTIONS
.TP
.B
\fB-2\fP
Create a 'deco 2' instead of a 'deco 1' architecture.
.TP
.B
\fB-b\fP\fIstrat\fP
Apply bipartitioning strategy \fIstrat\fP to compute the
recursive bipartition of the whole source graph into
smaller target subdomains.
.TP
.B
\fB-h\fP
Display some help.
.TP
.B
\fB-l\fP\fIlfile\fP
Only keep vertices the indices of which belong to the
space-separated list stored in \fIlfile\fP. This allows one
to create target architectures which can even be
disjoint subsets of a larger target architecture,
modeled as a graph.
.TP
.B
\fB-V\fP
Display program version and copyright.
.SH EXAMPLE
Create a decomposition-defined target architecture from a 2D regular
grid source graph of dimension 3 times 5, and save it, as a compiled
target architecture, under the \fBgzip\fP(1) format, to file 'm3x5.tgt.gz'. 
.PP
.nf
.fam C
    $ gmk_m2 3 5 | amk_grf | acpl - m3x5.tgt.gz

.fam T
.fi
Note that, in this precise case, it would be much preferable to use
directly the 'mesh2D' algorithmically-defined target architecture.
.PP
.nf
.fam C
    $ echo "mesh2D 3 5" > m3x5.tgt

.fam T
.fi
.SH SEE ALSO
\fBacpl\fP(1), \fBamk_ccc\fP(1), \fBatst\fP(1), \fBdgmap\fP(1), \fBgmap\fP(1), \fBgmtst\fP(1).
.PP
Scotch user's manual.
.SH AUTHOR
Francois Pellegrini <francois.pellegrini@labri.fr>


src/scotch/amk_grf.c
/* Copyright 2004,2007,2008,2010-2012,2014,2018,2019,2023 IPB, Universite de Bordeaux, INRIA & CNRS
**
** This file is part of the Scotch software package for static mapping,
** graph partitioning and sparse matrix ordering.
**
** This software is governed by the CeCILL-C license under French law
** and abiding by the rules of distribution of free software. You can
** use, modify and/or redistribute the software under the terms of the
** CeCILL-C license as circulated by CEA, CNRS and INRIA at the following
** URL: "http://www.cecill.info".
**
** As a counterpart to the access to the source code and rights to copy,
** modify and redistribute granted by the license, users are provided
** only with a limited warranty and the software's author, the holder of
** the economic rights, and the successive licensors have only limited
** liability.
**
** In this respect, the user's attention is drawn to the risks associated
** with loading, using, modifying and/or developing or reproducing the
** software by the user in light of its specific status of free software,
** that may mean that it is complicated to manipulate, and that also
** therefore means that it is reserved for developers and experienced
** professionals having in-depth computer knowledge. Users are therefore
** encouraged to load and test the software's suitability as regards
** their requirements in conditions enabling the security of their
** systems and/or data to be ensured and, more generally, to use and
** operate it in the same conditions as regards security.
**
** The fact that you are presently reading this means that you have had
** knowledge of the CeCILL-C license and that you accept its terms.
*/
/************************************************************/
/**                                                        **/
/**   NAME       : amk_grf.c                               **/
/**                                                        **/
/**   AUTHOR     : Francois PELLEGRINI                     **/
/**                                                        **/
/**   FUNCTION   : Creates the architecture description    **/
/**                file for any source graph.              **/
/**                                                        **/
/**   DATES      : # Version 3.0  : from : 06 jul 1995     **/
/**                                 to   : 02 oct 1995     **/
/**                # Version 3.1  : from : 26 mar 1996     **/
/**                                 to   : 26 mar 1996     **/
/**                # Version 3.2  : from : 23 apr 1997     **/
/**                                 to   : 03 jun 1998     **/
/**                # Version 3.3  : from : 15 may 1999     **/
/**                                 to   : 15 may 1999     **/
/**                # Version 3.4  : from : 03 feb 2000     **/
/**                                 to   : 03 feb 2000     **/
/**                # Version 4.0  : from : 11 dec 2001     **/
/**                                 to   : 17 mar 2005     **/
/**                # Version 5.0  : from : 23 dec 2007     **/
/**                                 to   : 16 mar 2008     **/
/**                # Version 5.1  : from : 11 dec 2008     **/
/**                                 to   : 17 jul 2011     **/
/**                # Version 6.0  : from : 01 jan 2012     **/
/**                                 to   : 24 sep 2019     **/
/**                # Version 7.0  : from : 21 jan 2023     **/
/**                                 to   : 21 jan 2023     **/
/**                                                        **/
/************************************************************/

/*
**  The defines and includes.
*/

#include "module.h"
#include "common.h"
#include "scotch.h"
#include "amk_grf.h"

/*
**  The static variables.
*/

static int                  C_fileNum = 0;        /* Number of file in arg list */
static File                 C_fileTab[C_FILENBR] = { /* File array              */
                              { FILEMODER },
                              { FILEMODEW },
                              { FILEMODER } };

static const char *         C_usageList[] = {     /* Usage */
  "amk_grf [<input source file> [<output target file>]] <options>",
  "  -2         : Create a 'deco 2' instead of a 'deco 1' architecture",
  "  -b<strat>  : Apply bipartitioning strategy <strat> (for 'deco 0' architectures)",
  "  -h         : Display this help",
  "  -l<file>   : Load vertex list from <file>",
  "  -V         : Print program version and copyright",
  "",
  "Default option set is : '-Bhf{move=1000}/((load0=load)|(load0=0))?x;'",
  NULL };

/******************************/
/*                            */
/* This is the main function. */
/*                            */
/******************************/

int
main (
int                         argc,
char *                      argv[])
{
  SCOTCH_Strat        bipastrat;                  /* Bipartitioning strategy                   */
  SCOTCH_Arch         archdat;                    /* Target (terminal) architecture            */
  SCOTCH_Graph        grafdat;                    /* Source graph to turn into architecture    */
  SCOTCH_Num          vertnbr;                    /* Number of vertices in graph               */
  SCOTCH_Num *        vlbltab;                    /* Pointer to vertex label array, if present */
  SCOTCH_Num          listnbr;                    /* Size of list array                        */
  SCOTCH_Num *        listtab;                    /* Pointer to list array                     */
  C_VertSort *        sorttab;                    /* Vertex label sort area                    */
  SCOTCH_Num          baseval;
  int                 flagval;                    /* Process flags                             */
  int                 i;

  errorProg ("amk_grf");

  if ((argc >= 2) && (argv[1][0] == '?')) {       /* If need for help */
    usagePrint (stdout, C_usageList);
    return     (EXIT_SUCCESS);
  }

  flagval = C_FLAGNONE;
  SCOTCH_stratInit (&bipastrat);

  fileBlockInit (C_fileTab, C_FILENBR);           /* Set default stream pointers */

  for (i = 1; i < argc; i ++) {                   /* Loop for all option codes                        */
    if ((argv[i][0] != '-') || (argv[i][1] == '\0') || (argv[i][1] == '.')) { /* If found a file name */
      if (C_fileNum < C_FILEARGNBR)               /* File name has been given                         */
        fileBlockName (C_fileTab, C_fileNum ++) = argv[i];
      else
        errorPrint ("main: too many file names given");
    }
    else {                                        /* If found an option name */
      switch (argv[i][1]) {
        case '2' :                                /* Type-2 architecture */
          flagval |= C_FLAGDECO2;
          break;
        case 'B' :                                /* Bipartitioning strategy */
        case 'b' :
          SCOTCH_stratExit (&bipastrat);
          SCOTCH_stratInit (&bipastrat);
          if ((SCOTCH_stratGraphBipart (&bipastrat, &argv[i][2])) != 0)
            errorPrint ("main: invalid bipartitioning strategy");
          break;
        case 'H' :                                /* Give the usage message */
        case 'h' :
          usagePrint (stdout, C_usageList);
          return     (EXIT_SUCCESS);
        case 'L' :                                /* Input vertex list */
        case 'l' :
          flagval |= C_FLAGVRTINP;
          if (argv[i][2] != '\0')
            C_filenamevrtinp = &argv[i][2];
          break;
        case 'V' :
          fprintf (stderr, "amk_grf, version " SCOTCH_VERSION_STRING "\n");
          fprintf (stderr, SCOTCH_COPYRIGHT_STRING "\n");
          fprintf (stderr, SCOTCH_LICENSE_STRING "\n");
          return  (EXIT_SUCCESS);
        default :
          errorPrint ("main: unprocessed option '%s'", argv[i]);
      }
    }
  }

  fileBlockOpen (C_fileTab, C_FILENBR);           /* Open all files */

  SCOTCH_graphInit (&grafdat);                    /* Create graph structure           */
  SCOTCH_graphLoad (&grafdat, C_filepntrgrfinp, -1, 0); /* Load source graph          */
  SCOTCH_graphData (&grafdat, &baseval, &vertnbr, NULL, NULL, NULL, /* Get graph data */
                    &vlbltab, NULL, NULL, NULL);

  listnbr = 0;                                    /* Initialize vertex list */
  listtab = NULL;
  if (flagval & C_FLAGVRTINP) {                   /* If list of vertices provided */
    SCOTCH_Num          listnum;

    if ((intLoad (C_filepntrvrtinp, &listnbr) != 1) || /* Read list size */
        (listnbr < 0)                               ||
        (listnbr > vertnbr))
      errorPrint ("main: bad list input (1)");

    if ((listtab = (SCOTCH_Num *) memAlloc (listnbr * sizeof (SCOTCH_Num))) == NULL)
      errorPrint ("main: out of memory (1)");

    for (listnum = 0; listnum < listnbr; listnum ++) { /* Read list data */
      if (intLoad (C_filepntrvrtinp, &listtab[listnum]) != 1)
        errorPrint ("main: bad list input (2)");
    }
    intSort1asc1 (listtab, listnbr);
    for (listnum = 0; listnum < listnbr - 1; listnum ++) { /* Search for duplicates */
      if (listtab[listnum] == listtab[listnum + 1])
        errorPrint ("main: duplicate list labels");
    }

    if (vlbltab != NULL) {                        /* If graph has vertex labels */
      SCOTCH_Num          vertnum;

      if ((sorttab = (C_VertSort *) memAlloc (vertnbr * sizeof (C_VertSort))) == NULL)
        errorPrint ("main: out of memory (2)");

      for (vertnum = 0; vertnum < vertnbr; vertnum ++) { /* Initialize sort area */
        sorttab[vertnum].vlblnum = vlbltab[vertnum];
        sorttab[vertnum].vertnum = vertnum;
      }
      intSort2asc1 (sorttab, vertnbr);            /* Sort by ascending labels */

      for (listnum = 0, vertnum = 0; listnum < listnbr; listnum ++) {  /* For all labels in list */
        while ((vertnum < vertnbr) && (sorttab[vertnum].vlblnum < listtab[listnum]))
          vertnum ++;                             /* Search vertex graph with corresponding label */
        if ((vertnum >= vertnbr) ||               /* If label not found                           */
            (sorttab[vertnum].vlblnum > listtab[listnum]))
          errorPrint ("main: list label '" SCOTCH_NUMSTRING "' not in graph", (SCOTCH_Num) listtab[listnum]);

        listtab[listnum] = sorttab[vertnum ++].vertnum; /* Replace label by number */
      }
      memFree (sorttab);                          /* Free sort area */
    }
  }

  SCOTCH_archInit  (&archdat);                    /* Initialize target architecture */
  if ((flagval & C_FLAGDECO2) != 0)
    SCOTCH_archBuild2 (&archdat, &grafdat, listnbr, listtab); /* Compute type-2 architecture */
  else
    SCOTCH_archBuild0 (&archdat, &grafdat, listnbr, listtab, &bipastrat); /* Compute type-1 architecture */
  SCOTCH_archSave  (&archdat, C_filepntrtgtout);  /* Write target architecture                           */

  fileBlockClose (C_fileTab, C_FILENBR);          /* Always close explicitely to end potential (un)compression tasks */

  SCOTCH_graphExit (&grafdat);                    /* Free target graph        */
  SCOTCH_archExit  (&archdat);                    /* Free target architecture */
  SCOTCH_stratExit (&bipastrat);                  /* Free strategy string     */
  if (listtab != NULL)                            /* If vertex list provided  */
    memFree (listtab);                            /* Free it                  */

  return (EXIT_SUCCESS);
}
